### ISSUE TITLE
Exception logging should not split a stack trace across lines.

### ISSUE THREAD

**Issue by airbyte-jenny** [airbytehq/airbyte#5269]
## Tell us about the problem you're trying to solve
The log files should keep a stack trace in a single line, so that they may be more sensibly grepped and ingested into standard logging systems such as splunk.  This makes it possible to have one ERROR line that reports the problem and the full stack trace, all together.  It also avoids the problem of multiple logging streams intermixing and confusing the issue.

## Describe the solution you’d like
The log4j formatter has a possible trick for replacing newlines with a different character, such as a pipe mark, so that multiline text can be displayed in the log file as a single line.  It's doable with a modern version of log4j and the right syntax in its config file.  This also paves the way for fully structured log output, such as pure json, which can be beneficial in the long-term as well.  

With log4j 4.2:
`<PatternLayout pattern="[%d] %-5p %m%n %throwable{separator(|)}"></PatternLayout>`

## Describe the alternative you’ve considered or used
Alternatives tend to be done at the log-ingest stage and are suboptimal because it's not always possible to piece back together the logging context once it's lost.

**Comment by airbyte-jenny** [comment]
Confirmed that this log4j2.xml pattern can work:
`%d{yyyy-MM-dd HH:mm:ss}{GMT+0} %p (%X{job_root}) %C{1}(%M):%L - %replace{%msg}{[\r\n]}{|} %throwable{separator(|)}%n`
specifically the %replace portion around the message and the %throwable portion to handle the stack trace.

Confirmed that this code can spit out where log4j2 is configured from:
```
  /**
   * Determine which config file is being used to configure the log system, and return its path within the classpath
   * @return A string with the path
   */
  public static String getLoggingConfigLocation() {
    try {
      LoggerContext logContext = LogManager.getContext();
      Field f = logContext.getClass().getDeclaredField("configuration");
      f.setAccessible(true);
      XmlConfiguration configuration = (XmlConfiguration) f.get(logContext);
      return configuration.getName();
    } catch (NoSuchFieldException | IllegalAccessException e) {
      return "Cannot read logger context configuration dynamically due to an API change.";
    }
  }
```

However, our various layers of docker and job submission and line gobbler result in some log lines not picking up the configuration easily from the airbyte-commons log4j2.xml file.  So there's more to be done to trace down all the methods by which logging config flows into the various subsystems.

**Comment by airbyte-jenny** [comment]
Rebuilding with specifically `SUB_BUILD=PLATFORM ./gradlew build` and `VERSION=dev docker-compose up -d` seems to cover all the lines except the stuff coming through the LineGobbler from child jobs.

**Comment by airbyte-jenny** [comment]
One thing this does mess up is the ascii-art "Airbyte" on server startup.
```
2021-08-13 21:09:44 INFO i.a.s.ServerApp(start):131 - {workspace_app_root=/tmp/workspace/server/logs} - |    ___    _      __          __|   /   |  (_)____/ /_  __  __/ /____|  / /| | / / ___/ __ \/ / / / __/ _ \| / ___ |/ / /  / /_/ / /_/ / /_/  __/|/_/  |_/_/_/  /_.___/\__, /\__/\___/|                    /____/|--------------------------------------| Now ready at http://localhost:8000/|--------------------------------------|Version: dev| 

```
That would need special handling if we want to keep the format looking right.

**Comment by airbyte-jenny** [comment]
LineGobbler output was the result of the postgres destination connector's main() definition:
```
public static void main(String[] args) throws Exception 
```
which was leaving it up to the jvm to handle the stack trace, instead of passing it through the LOGGER first.  Correcting this with a try-catch, correctly logging the error, and then using System.exit(1) to indicate a job failure, solves that logging issue (as well as, side effect, those errors are picked up by the default airbyte stream factory instead of by LineGobbler).  So the log4j2.xml was being correctly found in the classpath.  The connector just wasn't using the logging system, which could be a problem in several connectors.

All of this encourages the idea that we should get the logging config to be in the database instead of baked into connectors, but we will also need a pass through connector main()'s to ensure they actually use it.

**Comment by jrhizor** [comment]
For some platforms like splunk/datadog/stackdriver structured JSON logging may be useful for handling multi-line stack traces.

**Comment by airbyte-jenny** [comment]
Yes, json logging is one of the reasons I wanted to solve the broken-apart stack traces.  It's a powerful logging strategy when dealing with large production installations and rapid troubleshooting.  We'll need to audit connectors' main() methods to make sure they don't leave it up to the jvm to throw exceptions, but instead use the provided logging framework, for the json approach to be viable.  Some of them currently are missing exception handling.

**Comment by cgardens** [comment]
Based on the conversation we had in the platform meeting on this, I think we we want to allow this as something that can be set via an environment variable. This means for those using docker compose they can still leverage the more human readable version but for for those using cloud providers they can get the more log solution friendly version. Please correct me if I am mis remembering.
